File: novi-app/activations/src/main/java/org/novi/activations/DateTimeActivation.java
package org.novi.activations;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.novi.core.activations.BaseActivation;
import org.novi.core.exceptions.ConfigurationParseException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Map;

public class DateTimeActivation implements BaseActivation<DateTimeActivation.DateTimeActivationConfigRecord> {

    private DateTimeActivationConfigRecord configuration;

    private static final String DATE_FORMAT = "dd-MM-yyyy hh:mm";

    Logger logger = LoggerFactory.getLogger(DateTimeActivation.class);

    public record DateTimeActivationConfigRecord(Date startDateTime, Date endDateTime) {
    }

    @Override
    public DateTimeActivation configuration(String configuration) throws ConfigurationParseException {
        try {
            this.configuration = mapper().readValue(configuration, new TypeReference<>() {
            });
            return this;
        } catch (JsonProcessingException e) {
            throw new ConfigurationParseException(e);
        }
    }

    @Override
    public Boolean apply(String context) {
        try {
            ObjectMapper mapper = new ObjectMapper();
            JsonNode root = mapper.readTree(context);
            Map<String, Object> contextMap = mapper.treeToValue(root, Map.class);
            return evaluateFor(contextMap);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    public Boolean evaluateFor(Map<String, Object> context) {
        SimpleDateFormat df = new SimpleDateFormat(DATE_FORMAT);
        try {
            Date currentDateTime = df.parse((String) context.get(getName() + ".currentDateTime"));
            logger.debug("Checking for {} <= {} < {}", this.configuration().startDateTime, currentDateTime, this.configuration().endDateTime);
            return this.configuration().startDateTime().compareTo(currentDateTime) <= 0 && this.configuration().endDateTime().compareTo(currentDateTime) > 0;
        } catch (ParseException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public DateTimeActivationConfigRecord configuration() {
        return configuration;
    }

    public ObjectMapper mapper() {
        SimpleDateFormat df = new SimpleDateFormat(DATE_FORMAT);
        ObjectMapper mapper = new ObjectMapper();
        mapper.setDateFormat(df);
        return mapper;
    }
}

File: novi-app/activations/src/main/scala/org/novi/activations/dsl/DateTimeActivation.scala
package org.novi.activations.dsl

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.core.`type`.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import org.novi.activations.dsl.DateTimeActivation.DateRangeRecord
import org.novi.core.activations.BaseActivation
import org.novi.core.dsl.DslActivation
import org.slf4j.LoggerFactory

import java.text.{ParseException, SimpleDateFormat}
import java.util.{Date, Map}
import java.{lang, util}

class DateTimeActivation(configuration: String) extends DslActivation(configuration) {
  private val DATE_FORMAT = "dd-MM-yyyy hh:mm"
  val mapper = new ObjectMapper
  mapper.setDateFormat(new SimpleDateFormat(DATE_FORMAT))
  mapper.registerModule(DefaultScalaModule)
  private[activations] val logger = LoggerFactory.getLogger(classOf[DateTimeActivation])

  private val dateRange: DateRangeRecord = if (configuration != null) mapper.readValue(configuration, new TypeReference[DateRangeRecord]() {}) else DateRangeRecord(util.Date(), util.Date())

  def this() = {
    this(null)
  }

  override def apply(context: String): lang.Boolean = {
    try {
      val mapper = new ObjectMapper
      val root = mapper.readTree(context)
      val contextMap = mapper.treeToValue(root, classOf[util.Map[String, AnyRef]])
      val df = new SimpleDateFormat(DATE_FORMAT)
      try {
        val currentDateTime = df.parse(contextMap.get(getName + ".currentDateTime").asInstanceOf[String])
        logger.debug("Checking for {} <= {} < {}", dateRange.startDateTime, currentDateTime, dateRange.endDateTime)
        dateRange.startDateTime.compareTo(currentDateTime) <= 0 && dateRange.endDateTime.compareTo(currentDateTime) > 0
      } catch {
        case e: ParseException =>
          throw new RuntimeException(e)
      }
    } catch {
      case e: JsonProcessingException =>
        throw new RuntimeException(e)
    }
  }

  override def valueOf(configuration: String): BaseActivation[String] = DateTimeActivation(configuration)
}

object DateTimeActivation{
  case class DateRangeRecord(startDateTime: util.Date, endDateTime:util.Date)
}
